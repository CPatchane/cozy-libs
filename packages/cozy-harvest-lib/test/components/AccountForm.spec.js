/* eslint-env jest */
import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import {
  AccountForm,
  AccountFields,
  AccountField
} from 'components/AccountForm'

configure({ adapter: new Adapter() })

const fixtures = {
  fields: {
    username: {
      type: 'text'
    },
    passphrase: {
      type: 'password'
    }
  },
  sanitized: {
    username: {
      encrypted: false,
      required: true,
      type: 'text'
    },
    passphrase: {
      encrypted: true,
      required: true,
      type: 'password'
    }
  },
  account: {
    auth: {
      username: 'Toto',
      credentials_encrypted:
        'bmFjbGj8JQfzzfTQ2aGKTpI+HI9N8xKAQqPTPD6/84x5GyiHm2hdn7N6rO8cLTCnkdsnd2eFWJRf'
    }
  },
  oauth: {
    scope: 'test'
  }
}

const t = jest.fn()

describe('AccountForm', () => {
  beforeEach(() => {
    t.mockClear()
    t.mockImplementation(key => key)
  })

  it('should render', () => {
    const wrapper = shallow(<AccountForm fields={fixtures.fields} t={t} />)
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('should inject initial values from account', () => {
    const wrapper = shallow(
      <AccountForm
        initialValues={fixtures.account.auth}
        fields={fixtures.fields}
        t={t}
      />
    )
    expect(wrapper.props().initialValues).toEqual(fixtures.account.auth)
  })

  it('should redirect to OAuthForm', () => {
    const component = shallow(
      <AccountForm oauth={fixtures.oauth} t={t} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  describe('AccountFields', () => {
    it('should render', () => {
      const component = shallow(
        <AccountFields manifestFields={fixtures.fields} t={t} />
      ).getElement()
      expect(component).toMatchSnapshot()
    })

    it('should render encrypted fields with placeholder', () => {
      const component = shallow(
        <AccountFields manifestFields={fixtures.sanitized} t={t} />
      ).getElement()
      expect(component).toMatchSnapshot()
    })
  })

  describe('AccountField', () => {
    it('should render', () => {
      const wrapper = shallow(
        <AccountField {...fixtures.sanitized.username} name="username" t={t} />
      )
      const component = wrapper.dive().getElement()
      expect(component).toMatchSnapshot()
    })

    it('render password', () => {
      const wrapper = shallow(
        <AccountField
          {...fixtures.sanitized.passphrase}
          name="passphrase"
          t={t}
        />
      )
      const component = wrapper.dive().getElement()
      expect(component).toMatchSnapshot()
    })

    it('render dropdown', () => {
      const options = [
        { label: 'Option 1', value: 'option1' },
        { label: 'Option 2', value: 'option2' }
      ]

      const wrapper = shallow(
        <AccountField name="multiple" options={options} t={t} type="dropdown" />
      )
      const component = wrapper.dive().getElement()
      expect(component).toMatchSnapshot()
    })

    it('uses predefined label', () => {
      const wrapper = shallow(
        <AccountField
          label="login"
          name="username"
          required={true}
          type="text"
          t={t}
        />
      )
      expect(wrapper.props().label).toBe('fields.login.label')
    })

    it('ignores invalid predefined label', () => {
      const wrapper = shallow(
        <AccountField
          label="foo"
          name="username"
          required={true}
          type="text"
          t={t}
        />
      )
      expect(wrapper.props().label).toBe('fields.username.label')
    })
  })
})
