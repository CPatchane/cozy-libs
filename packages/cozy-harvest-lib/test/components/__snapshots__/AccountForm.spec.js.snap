// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm AccountField render a date field 1`] = `
<div
  className="o-field u-m-0"
>
  <Label
    className=""
    error={false}
    htmlFor=""
  >
    fields.date.label
  </Label>
  <Input
    autoComplete="off"
    disabled={false}
    error={false}
    fullwidth={true}
    id=""
    inputRef={[Function]}
    placeholder="fields.date.placeholder"
    size="medium"
    type="date"
    value=""
  />
</div>
`;

exports[`AccountForm AccountField render a dropdown field 1`] = `
<div
  className="o-field u-m-0"
>
  <Label
    className=""
    error={false}
    htmlFor=""
  >
    fields.multiple.label
  </Label>
  <Aware
    className="u-m-0"
    disabled={false}
    error={false}
    fieldProps={Object {}}
    fullwidth={true}
    id=""
    inputRef={[Function]}
    label="fields.multiple.label"
    name="multiple"
    options={
      Array [
        Object {
          "label": "Option 1",
          "value": "option1",
        },
        Object {
          "label": "Option 2",
          "value": "option2",
        },
      ]
    }
    placeholder="fields.multiple.placeholder"
    required={true}
    secondaryLabels={Object {}}
    side={null}
    size="medium"
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "legacy.fields.multiple.label",
            Object {
              "_": "multiple",
            },
          ],
          Array [
            "fields.multiple.label",
            Object {
              "_": "legacy.fields.multiple.label",
            },
          ],
          Array [
            "fields.multiple.placeholder",
            Object {
              "_": "",
            },
          ],
          Array [
            "accountForm.password.hide",
          ],
          Array [
            "accountForm.password.show",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "legacy.fields.multiple.label",
          },
          Object {
            "type": "return",
            "value": "fields.multiple.label",
          },
          Object {
            "type": "return",
            "value": "fields.multiple.placeholder",
          },
          Object {
            "type": "return",
            "value": "accountForm.password.hide",
          },
          Object {
            "type": "return",
            "value": "accountForm.password.show",
          },
        ],
      }
    }
    type="select"
  />
</div>
`;

exports[`AccountForm AccountField render a password field 1`] = `
<div
  className="o-field u-m-0"
>
  <Label
    className=""
    error={false}
    htmlFor=""
  >
    fields.passphrase.label
  </Label>
  <InputPassword
    autoComplete="new-password"
    disabled={false}
    error={false}
    fullwidth={true}
    hideLabel="accountForm.password.hide"
    id=""
    inputRef={[Function]}
    placeholder="fields.passphrase.placeholder"
    showLabel="accountForm.password.show"
    showVisibilityButton={true}
    size="medium"
    type="password"
    value=""
  />
</div>
`;

exports[`AccountForm AccountField should render 1`] = `
<div
  className="o-field u-m-0"
>
  <Label
    className=""
    error={false}
    htmlFor=""
  >
    fields.username.label
  </Label>
  <Input
    autoComplete="off"
    disabled={false}
    error={false}
    fullwidth={true}
    id=""
    inputRef={[Function]}
    placeholder="fields.username.placeholder"
    size="medium"
    type="text"
    value=""
  />
</div>
`;

exports[`AccountForm AccountFields should render 1`] = `
<div>
  <Field
    format={[Function]}
    name="username"
    parse={[Function]}
  >
    [Function]
  </Field>
  <Field
    format={[Function]}
    name="passphrase"
    parse={[Function]}
  >
    [Function]
  </Field>
</div>
`;

exports[`AccountForm AccountFields should render encrypted fields with placeholder 1`] = `
<div>
  <Field
    format={[Function]}
    name="username"
    parse={[Function]}
  >
    [Function]
  </Field>
  <Field
    format={[Function]}
    name="passphrase"
    parse={[Function]}
  >
    [Function]
  </Field>
</div>
`;

exports[`AccountForm should redirect to OAuthForm 1`] = `
<Wrapper
  oauth={
    Object {
      "scope": "test",
    }
  }
/>
`;

exports[`AccountForm should render 1`] = `
<div>
  <AccountFields
    container={null}
    initialValues={Object {}}
    manifestFields={
      Object {
        "passphrase": Object {
          "encrypted": true,
          "required": true,
          "type": "password",
        },
        "username": Object {
          "encrypted": false,
          "required": true,
          "role": "identifier",
          "type": "text",
        },
      }
    }
    onKeyUp={[Function]}
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "accountForm.submit.label",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "accountForm.submit.label",
          },
        ],
      }
    }
  />
  <DefaultButton
    className="u-mt-2 u-mb-1-half"
    disabled={true}
    extension="full"
    label="accountForm.submit.label"
    onClick={[Function]}
  />
</div>
`;
